File: frontend/lib/models/event.dart
class Event {
  final int id;
  final int organizerId;
  final String organizerName;
  final String title;
  final String description;
  final String category;
  final String venue;
  final String address;
  final String city;

  // Registration period — both required
  final DateTime registrationStartDateTime;
  final DateTime registrationEndDateTime;

  // Event schedule — start required, end optional
  final DateTime startDateTime;
  final DateTime? endDateTime; // null = no end time specified

  final double ticketPrice;
  final int capacity;
  final int availableTickets;
  final String status;
  final String? imageUrl;
  final DateTime createdAt;
  final DateTime updatedAt;

  Event({
    required this.id,
    required this.organizerId,
    required this.organizerName,
    required this.title,
    required this.description,
    required this.category,
    required this.venue,
    required this.address,
    required this.city,
    required this.registrationStartDateTime,
    required this.registrationEndDateTime,
    required this.startDateTime,
    this.endDateTime,
    required this.ticketPrice,
    required this.capacity,
    required this.availableTickets,
    required this.status,
    this.imageUrl,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Event.fromJson(Map<String, dynamic> json) {
    return Event(
      id: json['id'],
      organizerId: json['organizerId'],
      organizerName: json['organizerName'] ?? '',
      title: json['title'],
      description: json['description'],
      category: json['category'],
      venue: json['venue'],
      address: json['address'],
      city: json['city'],
      registrationStartDateTime:
          DateTime.parse(json['registrationStartDateTime']),
      registrationEndDateTime:
          DateTime.parse(json['registrationEndDateTime']),
      startDateTime: DateTime.parse(json['startDateTime']),
      endDateTime: json['endDateTime'] != null
          ? DateTime.parse(json['endDateTime'])
          : null,
      ticketPrice: (json['ticketPrice'] as num).toDouble(),
      capacity: json['capacity'],
      availableTickets: json['availableTickets'],
      status: json['status'],
      imageUrl: json['imageUrl'] as String?,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'organizerId': organizerId,
      'organizerName': organizerName,
      'title': title,
      'description': description,
      'category': category,
      'venue': venue,
      'address': address,
      'city': city,
      'registrationStartDateTime': registrationStartDateTime.toIso8601String(),
      'registrationEndDateTime': registrationEndDateTime.toIso8601String(),
      'startDateTime': startDateTime.toIso8601String(),
      'endDateTime': endDateTime?.toIso8601String(),
      'ticketPrice': ticketPrice,
      'capacity': capacity,
      'availableTickets': availableTickets,
      'status': status,
      'imageUrl': imageUrl,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

File: backend/src/main/java/com/eventmanagement/entity/Event.java
package com.eventmanagement.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "events")
public class Event {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "organizer_id", nullable = false)
    private User organizer;

    @NotBlank(message = "Title is required")
    @Size(min = 3, max = 200, message = "Title must be between 3 and 200 characters")
    @Column(nullable = false, length = 200)
    private String title;

    @NotBlank(message = "Description is required")
    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @NotBlank(message = "Category is required")
    @Column(nullable = false, length = 100)
    private String category;

    @NotBlank(message = "Venue is required")
    @Column(nullable = false, length = 200)
    private String venue;

    @NotBlank(message = "Address is required")
    @Column(nullable = false, length = 300)
    private String address;

    @NotBlank(message = "City is required")
    @Column(nullable = false, length = 100)
    private String city;

    // ── Registration period (REQUIRED) ────────────────────────────────────────
    @NotNull(message = "Registration start date/time is required")
    @Column(nullable = false)
    private LocalDateTime registrationStartDateTime;

    @NotNull(message = "Registration end date/time is required")
    @Column(nullable = false)
    private LocalDateTime registrationEndDateTime;

    // ── Event schedule ────────────────────────────────────────────────────────
    @NotNull(message = "Event start date/time is required")
    @Column(nullable = false)
    private LocalDateTime startDateTime;

    // endDateTime is intentionally nullable — event end time is optional
    @Column(nullable = true)
    private LocalDateTime endDateTime;

    @NotNull(message = "Ticket price is required")
    @DecimalMin(value = "0.00", message = "Ticket price must not be negative")
    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal ticketPrice;

    @NotNull(message = "Capacity is required")
    @Min(value = 1, message = "Capacity must be greater than zero")
    @Column(nullable = false)
    private Integer capacity;

    @Column(nullable = false)
    private Integer availableTickets;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private EventStatus status = EventStatus.PUBLISHED;

    @Column(length = 512)
    private String imageUrl;

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(nullable = false)
    private LocalDateTime updatedAt;

    public Event() {
    }

    // ── Getters and Setters ───────────────────────────────────────────────────

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public User getOrganizer() { return organizer; }
    public void setOrganizer(User organizer) { this.organizer = organizer; }

    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public String getCategory() { return category; }
    public void setCategory(String category) { this.category = category; }

    public String getVenue() { return venue; }
    public void setVenue(String venue) { this.venue = venue; }

    public String getAddress() { return address; }
    public void setAddress(String address) { this.address = address; }

    public String getCity() { return city; }
    public void setCity(String city) { this.city = city; }

    public LocalDateTime getRegistrationStartDateTime() { return registrationStartDateTime; }
    public void setRegistrationStartDateTime(LocalDateTime registrationStartDateTime) {
        this.registrationStartDateTime = registrationStartDateTime;
    }

    public LocalDateTime getRegistrationEndDateTime() { return registrationEndDateTime; }
    public void setRegistrationEndDateTime(LocalDateTime registrationEndDateTime) {
        this.registrationEndDateTime = registrationEndDateTime;
    }

    public LocalDateTime getStartDateTime() { return startDateTime; }
    public void setStartDateTime(LocalDateTime startDateTime) { this.startDateTime = startDateTime; }

    public LocalDateTime getEndDateTime() { return endDateTime; }
    public void setEndDateTime(LocalDateTime endDateTime) { this.endDateTime = endDateTime; }

    public BigDecimal getTicketPrice() { return ticketPrice; }
    public void setTicketPrice(BigDecimal ticketPrice) { this.ticketPrice = ticketPrice; }

    public Integer getCapacity() { return capacity; }
    public void setCapacity(Integer capacity) { this.capacity = capacity; }

    public Integer getAvailableTickets() { return availableTickets; }
    public void setAvailableTickets(Integer availableTickets) { this.availableTickets = availableTickets; }

    public EventStatus getStatus() { return status; }
    public void setStatus(EventStatus status) { this.status = status; }

    public String getImageUrl() { return imageUrl; }
    public void setImageUrl(String imageUrl) { this.imageUrl = imageUrl; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
}
